File: missingNumberinArray.py
# Time Complexity : O( log N)
# Space Complexity : O(N)
# Did this code successfully run on Leetcode :no, executed in PyCharm
# Any problem you faced while coding this : no
def search(arr, size):
    low = 0
    high = size - 1
    mid = 0
    while high > low + 1:
        mid = low + (high - low) // 2
        midIndexDiff = arr[mid] - mid
        lowIndexDiff = arr[low] - low
        highIndexDiff = arr[high] - high
        if midIndexDiff != lowIndexDiff:
            high = mid
        elif midIndexDiff != highIndexDiff:
            low = mid
    return (arr[low] + arr[high]) // 2


arr = [1, 2, 3, 5]
size = len(arr)

print("Missing number:", search(arr, size))

File: Missing_number_in_array.cpp

class Solution{
  public:
    int missingNumber(vector<int>& a, int n) {
        // Your code goes here
        sort(a.begin(), a.end());
        int low=0;
        int high=a.size()-1;
        if(a[0]!=1)return 1;
        if(a[a.size()-1]!=n)return n;
        while(low<=high){
            int mid = low + (high-low)/2;
            if(a[mid]==mid+2){
                if(mid>0 && a[mid]-a[mid-1]==2){
                    return a[mid]-1; 
                }
                else{
                    high = mid-1;
                }
            }
            
            else {
                if(mid<a.size())
                low = mid+1;
            }
        }
    
        return -1;
    }
};
